test/scale/scale_tester.go

```diff
@@ -0,0 +1,292 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scale
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/onsi/ginkgo"
+	scale "k8s.io/api/autoscaling/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"sigs.k8s.io/azuredisk-csi-driver/test/e2e/driver"
+	testtypes "sigs.k8s.io/azuredisk-csi-driver/test/types"
+	nodeutil "sigs.k8s.io/azuredisk-csi-driver/test/utils/node"
+	podutil "sigs.k8s.io/azuredisk-csi-driver/test/utils/pod"
+	volutil "sigs.k8s.io/azuredisk-csi-driver/test/utils/volume"
+)
+
+type PodSchedulingWithPVScaleTest struct {
+	CSIDriver              driver.DynamicPVTestDriver
+	Pod                    testtypes.PodDetails
+	Replicas               int
+	StorageClassParameters map[string]string
+}
+
+type PodSchedulingOnFailoverScaleTest struct {
+	CSIDriver              driver.DynamicPVTestDriver
+	Pod                    testtypes.PodDetails
+	PodCount               int
+	Replicas               int
+	StorageClassParameters map[string]string
+}
+
+func (t *PodSchedulingWithPVScaleTest) Run(client clientset.Interface, namespace *v1.Namespace, schedulerName string) {
+	totalNumberOfPods := t.Replicas
+	var oneMinPodsRunning,
+		twoMinPodsRunning,
+		threeMinPodsRunning,
+		fiveMinPodsRunning,
+		tenMinPodsRunning = totalNumberOfPods, totalNumberOfPods, totalNumberOfPods, totalNumberOfPods, totalNumberOfPods
+	var oneMinPodsDeleted,
+		twoMinPodsDeleted,
+		threeMinPodsDeleted,
+		fiveMinPodsDeleted,
+		tenMinPodsDeleted = totalNumberOfPods, totalNumberOfPods, totalNumberOfPods, totalNumberOfPods, totalNumberOfPods
+
+	startScaleOut := time.Now()
+	tStorageClass, storageCleanup := t.Pod.CreateStorageClass(client, namespace, t.CSIDriver, t.StorageClassParameters)
+	defer storageCleanup()
+	statefulSet, cleanupStatefulSet := t.Pod.SetupStatefulset(client, namespace, t.CSIDriver, schedulerName, t.Replicas, t.StorageClassParameters, &tStorageClass)
+	for i := range cleanupStatefulSet {
+		i := i
+		defer cleanupStatefulSet[i]()
+	}
+	statefulSet.CreateWithoutWaiting()
+	numberOfRunningPods := 0
+	ticker := time.NewTicker(time.Minute)
+	tickerStop := make(chan bool)
+	go func() {
+		time.Sleep(11 * time.Minute)
```

Comment:
I do think we still need the goroutine and the channel to wait for the whole timeout period. On each iteration `select` is going to block the loop until wither `<-time.After()` or `ticker.C` return a value on their channels. Since `ticker.C` is the one sending a value through the channel every minute, it is always going to be the selected option, and on the next iteration we will end up creating a new channel for `<-time.After()` and wait on the next tick and so on... Thus, we will create an infinite loop that will not timeout if we rely on `<-time.After()` in select.

Reply:
But we should definitely update and create a make(chan struct{}) instead of bool.

Reply:
`time.After` doesn't have to be called in a `select` `case` statement. It can be assigned to a variable, e.g.:

```go
ticker := time.NewTicker(time.Minute)
timeout := time.After(*scaleTestTimeout)

for {
  select {
    case <-ticker:
      ...
    case <-timeout:
      ...
  }
}
```

Reply:
Yep, this would certainly work, and I don't know this didn't even cross my mind :D